src/Dapr.Actors/Runtime/ActorManager.cs

```diff
@@ -199,25 +202,86 @@ async Task<byte[]> RequestFunc(Actor actor, CancellationToken ct)
             return this.DispatchInternalAsync(actorId, this.timerMethodContext, RequestFunc, cancellationToken);
         }
 
-        internal async Task ActivateActor(ActorId actorId)
+        internal async Task ActivateActorAsync(ActorId actorId)
         {
             // An actor is activated by "Dapr" runtime when a call is to be made for an actor.
-            var actor = this.actorService.CreateActor(actorId);
-            await actor.OnActivateInternalAsync();
+            var state = await this.CreateActorAsync(actorId);
+
+            try
+            {
+                await state.Actor.OnActivateInternalAsync();
+            }
+            catch
+            {
+                // Ensure we don't leak resources if user-code throws during activation.
+                await DeleteActorAsync(state);
+                throw;
+            }
 
             // Add actor to activeActors only after OnActivate succeeds (user code can throw error from its override of Activate method.)
-            // Always add the new instance.
-            this.activeActors.AddOrUpdate(actorId, actor, (key, oldValue) => actor);
+            //
+            // In theory the Dapr runtime protects us from double-activation - there's no case
+            // where we *expect* to see the *update* code path taken. However it's a possiblity and
+            // we should handle it.
+            //
+            // The policy we have chosen is to always keep the registered instance if we hit a double-activation
+            // so that means we have to destroy the 'new' instance.
+            var current = this.activeActors.AddOrUpdate(actorId, state, (key, oldValue) => oldValue);
```

Comment:
Why don't we check if the actor is active and activate it/add it only if it does not exist?

Reply:
That approach fundamentally has a race condition if the existing actor is being deactivated at the same time. Concurrent dictionary doesn't provide the guarantee that you can do that atomically.